neurokit2/ecg/ecg_findpeaks.py

```diff
@@ -1036,65 +1036,53 @@ def _ecg_findpeaks_MWA(signal, window_size):
 
 
 def _ecg_findpeaks_peakdetect(detection, sampling_rate=1000):
-    """From https://github.com/berndporr/py-ecg-detectors/"""
-    min_distance = int(0.25 * sampling_rate)
-
-    signal_peaks = [0]
-    noise_peaks = []
-
-    SPKI = 0.0
-    NPKI = 0.0
+    """Based on https://github.com/berndporr/py-ecg-detectors/
 
-    threshold_I1 = 0.0
-    threshold_I2 = 0.0
+    Optimized for vectorized computation.
 
-    RR_missed = 0
-    index = 0
-    indexes = []
+    """
+    min_peak_distance = int(0.3 * sampling_rate)
+    min_missed_distance = int(0.25 * sampling_rate)
 
-    missed_peaks = []
-    peaks = []
+    signal_peaks = []
 
-    for i in range(len(detection)):  # pylint: disable=R1702,C0200
+    SPKI = 0.0
+    NPKI = 0.0
 
-        # pylint: disable=R1716
-        if i > 0 and i < len(detection) - 1 and detection[i - 1] < detection[i] and detection[i + 1] < detection[i]:
-            peak = i
-            peaks.append(peak)  # pylint: disable=R1716
-            if detection[peak] > threshold_I1 and (peak - signal_peaks[-1]) > 0.3 * sampling_rate:
-
-                signal_peaks.append(peak)
-                indexes.append(index)
-                SPKI = 0.125 * detection[signal_peaks[-1]] + 0.875 * SPKI
-                if RR_missed != 0 and signal_peaks[-1] - signal_peaks[-2] > RR_missed:
-                    missed_section_peaks = peaks[indexes[-2] + 1 : indexes[-1]]
-                    missed_section_peaks2 = []
-                    for missed_peak in missed_section_peaks:
-                        if missed_peak - signal_peaks[-2] > min_distance:
-                            if signal_peaks[-1] - missed_peak > min_distance:
-                                if detection[missed_peak] > threshold_I2:
-                                    missed_section_peaks2.append(missed_peak)
-
-                    if missed_section_peaks2:
-                        missed_peak = missed_section_peaks2[np.argmax(detection[missed_section_peaks2])]
-                        missed_peaks.append(missed_peak)
-                        signal_peaks.append(signal_peaks[-1])
-                        signal_peaks[-2] = missed_peak
+    last_peak = 0
+    last_index = -1
 
-            else:
-                noise_peaks.append(peak)
-                NPKI = 0.125 * detection[noise_peaks[-1]] + 0.875 * NPKI
+    # TODO: Using plateau_size=(1,1) here avoids detecting flat peaks and
+    # maintains original py-ecg-detectors behaviour, but is not obviously
+    # correct. Should we remove that setting to find also flat peaks?
```

Comment:
> Should we remove that setting to find also flat peaks?

@JanCBrammer any thoughts?

Reply:
I looked at both the Pan-Tompkins and Kalidas-Tamil papers (that describe the algorithms this function is used for) but neither is particularly specific about this detail. Pan-Tompkins just says:

> A peak is a local maximum determined by observing when the
> signal changes direction within a predefined time interval.

In this implementation that "predefined time interval" is just three subsequent samples.

I looked at the data I have, and didn't spot any cases where flat peaks would actually be a problem. It looks like the only situations where flat peak detection would make a difference are things like artefacts where the signal gets saturated at maximum recording amplitude. There's no point in even trying to detect peaks from such data, so I'm not sure if there's any compelling reason to consider changing the `plateau_size` setting. And if we did want to change it, the setting should probably not exceed the length of typical QRS complexes to avoid marking saturation artefacts as peaks.

To summarize, something like `plateau_size=(1, max(1, sampling_rate // 10))` might be reasonable, though not necessarily worth actually implementing.

Reply:
I changed the TODO to a NOTE comment in https://github.com/neuropsychology/NeuroKit/pull/450/commits/5a1d82919878e5c781375f18869df8b205eec2c0 based on the above observations.